## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

---
### View the column names. (4 points)

In [4]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

---
### Use the describe function. (4 points)

In [5]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

---
### Correctly identify the category with the most entries. (4 points)

In [12]:
# Use this space to do any additional research
# and familiarize yourself with the data.
category_counts = df["category"].value_counts()
print(category_counts)

max_category = category_counts.idxmax()
print(f"\nThe category with the most entries is: {max_category}")

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

The category with the most entries is: consumables


---
### For the category with the most entries, correctly identify the subcategory with the most entries. (5 points)

In [15]:
max_category_sub_counts = df[df["category"] == max_category]["subcategory"].value_counts()
print(max_category_sub_counts)

max_category_max_sub = max_category_sub_counts.idxmax()
print(f"\nThe sub-category with the most entries in the most common category is: {max_category_max_sub}")

subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

The sub-category with the most entries in the most common category is: bathroom supplies


---
### Correctly identify the 5 clients with the most entries in the data. (5 points)

In [21]:
# Which five clients had the most entries in the data?
most_clients = df["client_id"].value_counts().head(5)
print("The clients with the most entries by client_id are:")
for client_id, count in most_clients.items():
    print(f"Client ID: {client_id}, Entries: {count}")

The clients with the most entries by client_id are:
Client ID: 33615, Entries: 220
Client ID: 66037, Entries: 211
Client ID: 46820, Entries: 209
Client ID: 38378, Entries: 207
Client ID: 24741, Entries: 207


---
### Store the client ids of those top 5 clients in a list. (4 points)

In [22]:
# Store the client ids of those top 5 clients in a list.
most_client_ids = most_clients.index.tolist()
print(f"the client ids of the top 5 clients are: {most_client_ids}")

the client ids of the top 5 clients are: [33615, 66037, 46820, 38378, 24741]


---
### Display the total units (the qty column) that the client with the most entries ordered. (4 points)

In [23]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = most_clients.idxmax()
top_client_units = df.loc[df["client_id"] == top_client_id, "qty"].sum()
print(f"The client ({top_client_id}) with the most entries ordered a total of {top_client_units} units.")

The client (33615) with the most entries ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

---
### Create a column that calculates the subtotal for each line using the unit_price and the qty. (6 points)

In [26]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["unit_price"] * df["qty"]

print("Preview of the new 'subtotal' column:")
df[["unit_price", "qty", "subtotal"]]

Preview of the new 'subtotal' column:


unit_price   qty   subtotal
0         1096.80   105  115164.00
1           24.95    21     523.95
2           13.52    39     527.28
3           36.42    29    1056.18
4          195.10    20    3902.00
...           ...   ...        ...
54634       83.13    33    2743.29
54635      206.59    47    9709.73
54636       65.66   475   31188.50
54637        1.48   112     165.76
54638        3.01  1031    3103.31

[54639 rows x 3 columns]

---
### Create a column for shipping price. Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under. (6 points)

In [35]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
standard_ship_rate = 10.00
discount_ship_rate = 7.00

# Create a column for the total_weight of each order.
df["total_weight"] = df["unit_weight"] * df["qty"]

# Set the shipping rate for all orders.
df["shipping_rate"] = standard_ship_rate

# Set the shipping rate for orders over 50 pounds.
df.loc[df["total_weight"] > 50, "shipping_rate"] = discount_ship_rate

# Calculate the shipping price for each line.
df["shipping_price"] = df["shipping_rate"] * df["total_weight"]

print("Preview of the new columns and their input columns:")
df[["unit_weight", "qty", "total_weight", "shipping_rate", "shipping_price"]]


Preview of the new columns and their input columns:


unit_weight   qty  total_weight  shipping_rate  shipping_price
0             7.50   105        787.50            7.0         5512.50
1             1.49    21         31.29           10.0          312.90
2             1.68    39         65.52            7.0          458.64
3             1.23    29         35.67           10.0          356.70
4            46.43    20        928.60            7.0         6500.20
...            ...   ...           ...            ...             ...
54634         2.25    33         74.25            7.0          519.75
54635        11.70    47        549.90            7.0         3849.30
54636         4.16   475       1976.00            7.0        13832.00
54637        18.04   112       2020.48            7.0        14143.36
54638         2.07  1031       2134.17            7.0        14939.19

[54639 rows x 5 columns]

---
### Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%. (6 points)

In [37]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = round((df["subtotal"] + df["shipping_price"]) * 1.0925, 2)

print("Preview of the new 'total_price' column and it's input columns:")
df[["subtotal", "shipping_price", "total_price"]]

Preview of the new 'total_price' column and it's input columns:


subtotal  shipping_price  total_price
0      115164.00         5512.50    131839.08
1         523.95          312.90       914.26
2         527.28          458.64      1077.12
3        1056.18          356.70      1543.57
4        3902.00         6500.20     11364.40
...          ...             ...          ...
54634    2743.29          519.75      3564.87
54635    9709.73         3849.30     14813.24
54636   31188.50        13832.00     49184.90
54637     165.76        14143.36     15632.71
54638    3103.31        14939.19     19711.43

[54639 rows x 3 columns]

---
### Create a column for the cost of each line using unit cost, qty, and shipping price (assume the shipping cost is exactly what is charged to the client). (6 points)

In [38]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["order_cost"] = df["unit_cost"] * df["qty"] + df["shipping_price"]

print("Preview of the new 'order_cost' column and it's input columns:")
df[["unit_cost", "qty", "shipping_price", "order_cost"]]


Preview of the new 'order_cost' column and it's input columns:


unit_cost   qty  shipping_price  order_cost
0         762.71   105         5512.50    85597.05
1          15.09    21          312.90      629.79
2           7.86    39          458.64      765.18
3          24.85    29          356.70     1077.35
4         108.17    20         6500.20     8663.60
...          ...   ...             ...         ...
54634      51.60    33          519.75     2222.55
54635     175.46    47         3849.30    12095.92
54636      57.31   475        13832.00    41054.25
54637       1.22   112        14143.36    14280.00
54638       1.61  1031        14939.19    16599.10

[54639 rows x 4 columns]

---
### Create a column for the profit of each line using line cost and line price. (6 points)

In [40]:
# Create a column for the profit of each line using line cost and line price
df["order_profit"] = df["total_price"] - df["order_cost"]

print("Preview of the new 'order_profit' column and it's input columns:")
df[["total_price", "order_cost", "order_profit"]]


Preview of the new 'order_profit' column and it's input columns:


total_price  order_cost  order_profit
0        131839.08    85597.05      46242.03
1           914.26      629.79        284.47
2          1077.12      765.18        311.94
3          1543.57     1077.35        466.22
4         11364.40     8663.60       2700.80
...            ...         ...           ...
54634      3564.87     2222.55       1342.32
54635     14813.24    12095.92       2717.32
54636     49184.90    41054.25       8130.65
54637     15632.71    14280.00       1352.71
54638     19711.43    16599.10       3112.33

[54639 rows x 3 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [145]:
# Check your work using the totals above
orders_to_check = [2742071, 2173913, 6128929]
# make a for loop that prints the order amount after recalculating the total price
for order in orders_to_check:
    order_df = df.loc[df["order_id"] == order]
    recalc_total = order_df["line_subtotal"].sum() + order_df["shipping_price"].sum()
    print(f"Order {order} total: {recalc_total * 1.0925:.2f}")

Order 2742071 total: 153250.30
Order 2173913 total: 163237.07
Order 6128929 total: 925642.41


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [146]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in most_client_ids:
    client_df = df.loc[df["client_id"] == client]
    client_total = client_df["total_price"].sum()
    print(f"Client {client} total: {client_total:.2f}")


Client 33615 total: 8428398.45
Client 66037 total: 10296153.42
Client 46820 total: 9788108.50
Client 38378 total: 12943277.51
Client 24741 total: 82296551.30


In [205]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
summary_df = pd.DataFrame({
    "total_units": df.groupby("client_id")["qty"].sum(),
    "total_shipping": df.groupby("client_id")["shipping_price"].sum(),
    "total_revenue": df.groupby("client_id")["total_price"].sum(),
    "total_cost": df.groupby("client_id")["line_cost"].sum(),
    "total_profit": df.groupby("client_id")["line_profit"].sum()
})
summary_df = summary_df.loc[most_client_ids]
summary_df.reset_index(inplace=True)
summary_df


client_id  total_units  total_shipping  total_revenue   total_cost  \
0      33615        64313      1875749.10     8428398.45   6222078.12   
1      66037        43018      1428688.37    10296153.42   7038019.50   
2      46820        75768      1642011.00     9788108.50   7047753.80   
3      38378        73667      3463072.47    12943277.51   9668338.05   
4      24741       239862      5151765.79    82296551.30  45714217.13   

   total_profit  
0    2206320.33  
1    3258133.92  
2    2740354.70  
3    3274939.46  
4   36582334.17

In [206]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.rename(lambda x: x.title().replace("_", " "), axis=1, inplace=True)
summary_df.rename(columns={"Client Id": "Client ID"}, inplace=True)
summary_df.rename(columns={"Total Units": "Units"}, inplace=True)

# Define the money columns.
money_columns = [col for col in summary_df.columns if 'Total' in col]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return f"{value / 1000000:.6f}"

# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
summary_df.rename(columns={col: f"{col} (millions)" for col in money_columns}, inplace=True)
summary_df



Client ID   Units Total Shipping (millions) Total Revenue (millions)  \
0      33615   64313                  1.875749                 8.428398   
1      66037   43018                  1.428688                10.296153   
2      46820   75768                  1.642011                 9.788108   
3      38378   73667                  3.463072                12.943278   
4      24741  239862                  5.151766                82.296551   

  Total Cost (millions) Total Profit (millions)  
0              6.222078                2.206320  
1              7.038019                3.258134  
2              7.047754                2.740355  
3              9.668338                3.274939  
4             45.714217               36.582334

In [207]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df_sorted = summary_df.sort_values("Total Profit (millions)", ascending=False)
summary_df_sorted

Client ID   Units Total Shipping (millions) Total Revenue (millions)  \
4      24741  239862                  5.151766                82.296551   
3      38378   73667                  3.463072                12.943278   
1      66037   43018                  1.428688                10.296153   
2      46820   75768                  1.642011                 9.788108   
0      33615   64313                  1.875749                 8.428398   

  Total Cost (millions) Total Profit (millions)  
4             45.714217               36.582334  
3              9.668338                3.274939  
1              7.038019                3.258134  
2              7.047754                2.740355  
0              6.222078                2.206320

In [ ]:
# We have found that the biggest indicator of profit is the total units purchased.
# Furthermore, client with the most units purchased also had the lowest cost per unit,
# suggesting that discounts may be applied as total unit purchase increased.